source/common/common/utility.cc

```diff
@@ -20,8 +21,116 @@
+  if (item == cached_time.formatted.end()) {
+    const time_t current_time = std::chrono::system_clock::to_time_t(time);
+
+    // Build a new formatted format string at current time.
+    CachedTime::Formatted formatted;
+    if (subseconds_.empty()) {
+      formatted.str = fromTime(current_time);
+    } else {
+      formatted.str = fromTimeAndPrepareSubsecondOffsets(current_time, formatted.subsecond_offsets);
+      cached_time.seconds_length = fmt::FormatInt(epoch_time_seconds.count()).str().size();
+    }
+
+    // Stamp the formatted string using the current epoch time in seconds, and then cache it in.
+    formatted.epoch_time_seconds = epoch_time_seconds;
+    cached_time.formatted.emplace(std::make_pair(format_string_, formatted));
+  }
+
+  const auto& formatted = cached_time.formatted.at(format_string_);
+  if (subseconds_.empty()) {
+    return formatted.str;
+  }
+
+  ASSERT(subseconds_.size() == formatted.subsecond_offsets.size());
+
+  // Copy the current cached formatted format string, then replace its subseconds part using the
+  // prepared subseconds offsets.
+  std::string formatted_str = formatted.str;
+  const std::string nanoseconds = fmt::FormatInt(epoch_time_ns.count()).str();
+  for (size_t i = 0; i < subseconds_.size(); ++i) {
+    const auto& subsecond = subseconds_.at(i);
+    formatted_str.replace(subsecond.position_ + formatted.subsecond_offsets.at(i), subsecond.width_,
+                          nanoseconds.substr(cached_time.seconds_length, subsecond.width_));
+  }
+
+  ASSERT(formatted_str.size() == formatted.str.size());
+  return formatted_str;
+}
+
+std::string DateFormatter::parse(const std::string& format_string) {
+  std::string new_format_string = format_string;
+  std::smatch matched;
+  size_t step = 0;
+  while (regex_search(new_format_string, matched, SubsecondConstants::get().PATTERN)) {
+    const std::string& width_specifier = matched[1];
+    const size_t width = width_specifier.empty() ? SubsecondConstants::get().PLACEHOLDER.size()
+                                                 : width_specifier.at(0) - '0';
+    new_format_string.replace(matched.position(), matched.length(),
+                              SubsecondConstants::get().PLACEHOLDER.substr(0, width));
+
+    ASSERT(step < new_format_string.size());
+
+    // This records matched position, the width of current subsecond pattern, and also the string
+    // segment before the matched position. These values will be used later at data path.
+    SubsecondSpecifier subsecond(matched.position(), width,
+                                 new_format_string.substr(step, matched.position() - step));
+    subseconds_.emplace_back(subsecond);
+
+    step = subsecond.position_ + subsecond.width_;
+  }
+
+  // To capture the segment after the last subsecond pattern of a format string. E.g.
+  // %3f-this-is-the-last-%s-segment-%Y-until-this.
+  if (step < new_format_string.size()) {
+    last_segment_ = new_format_string.substr(step);
```

Comment:
Can you create a dummy zero length subsecond entry to allow this to be treated more uniformly later?